zenoh/src/net/runtime/orchestrator.rs

```diff
@@ -291,24 +291,119 @@ impl Runtime {
+        //
+        //      "tcp/192.168.0.1:7447#retry=1;tm=1000;incfactor=2;exit_on_fail=false"
+        //              but it not consistent with global config
+        //
+        let config = endpoint.config();
+        if let Some(val) = config.get("retry_count") {
+            res.count = zparse_default!(val, res.count);
+        }
+        if let Some(val) = config.get("retry_timeout_ms") {
+            res.timeout_ms = zparse_default!(val, res.timeout_ms);
+        }
+        if let Some(val) = config.get("retry_timeout_increase_factor") {
+            res.timeout_increase_factor = zparse_default!(val, res.timeout_increase_factor);
+        }
+        if let Some(val) = config.get("retry_exit_on_fail") {
+            //TODO(sashacmc): rewrite to use separate structure or macro
+            res.exit_on_fail = ModeDependentValue::Unique(zparse_default!(
+                val,
+                *res.exit_on_fail.get(self.whatami()).unwrap_or(&false)
+            ));
+        }
+        res
+    }
+
     async fn bind_listeners(&self, listeners: &[EndPoint]) -> ZResult<()> {
         for listener in listeners {
             let endpoint = listener.clone();
-            match self.manager().add_listener(endpoint).await {
-                Ok(listener) => log::debug!("Listener added: {}", listener),
-                Err(err) => {
-                    log::error!("Unable to open listener {}: {}", listener, err);
-                    return Err(err);
+            let retry_config = self.get_retry_config(&endpoint, true);
+            log::info!(
+                "Retry!!!!!!!!!!!!!! {}: {}",
+                retry_config.count,
+                retry_config.timeout_ms
+            );
+            if retry_config.count > 0 {
+                self.spawn_add_listener(endpoint, retry_config).await?
+            } else {
+                self.add_listener(endpoint).await?
+            }
+        }
+
+        self.print_locators();
+
+        Ok(())
+    }
+
+    async fn spawn_add_listener(
+        &self,
+        listener: EndPoint,
+        retry_config: zenoh_config::ConnectionRetryConf,
+    ) -> ZResult<()> {
+        let this = self.clone();
+        self.spawn(async move {
+            let mut delay = retry_config.timeout_ms;
+            let mut success = false;
+            for _ in 0..retry_config.count {
+                if this.add_listener(listener.clone()).await.is_ok() {
+                    success = true;
+                    break;
+                }
+                async_std::task::sleep(Duration::from_millis(delay)).await;
+                delay = (delay as f32 * retry_config.timeout_increase_factor) as u64;
+            }
+            if success {
+                this.print_locators();
+            } else {
+                if *retry_config.exit_on_fail.get(this.whatami()).unwrap() {
+                    //TODO(sashacmc): how to exit, this is don't work?
+                    panic!("Unable to connect");
```

Comment:
How correctly terminate a program from the thread? Do we already have it somewhere?